### PR TITLE
Fix wrong handling of `grad_scale` & `found_inf` in fused optimizers

### LINKED ISSUE #95781: Fused AdamW causes NaN loss
### 🐛 Describe the bug

There already has been an extended discussion of this issue over on the nanoGPT repository:
- karpathy/nanoGPT#167

I have been encouraged to submit a separate bug report here, since the issue seems to lie with pytorch. I have prepared a preconfigured [fork of nanoGPT](https://github.com/oddlama/nanoGPT_nan) where I provided my configuration and part of my dataset which causes the issue to appear. It now immediately produces NaNs after the first training step, but only if fused adam is in use.

The settings are basically the nanoGPT shakespeare configuration, but using a blocksize of 343 and vocab size of 2006. The data is quite sparse (only 1,3% is not 0, which significantly accelerates getting to the issue). I've included 1000 batches of my own real data so you will have the same conditions as I do. I'm training on a single 2080Ti, please look at a diff of the latest commit in this repository to see what I changed exactly - it's not much.

```bash
> git clone https://github.com/oddlama/nanoGPT_nan
> cd nanoGPT_nan
```
```bash
> python train.py config/nan.py --allow_fused=True
[...]
number of parameters: 11.39M
using fused AdamW: True
compiling the model... (takes a ~minute)
step 0: train loss 6.6010, val loss 6.6009
[2023-03-01 15:19:43,485] torch._inductor.utils: [WARNING] using triton random, expect difference from eager
/projects/venv/lib/python3.10/site-packages/torch/_functorch/aot_autograd.py:1251: UserWarning: Your compiler for AOTAutograd is returning a a function that doesn't take boxed arguments. Please wrap it with functorch.compile.make_boxed_func or handle the boxed arguments yourself. See https://github.com/pytorch/pytorch/pull/83137#issuecomment-1211320670 for rationale.
  warnings.warn(
iter 0: loss 6.6793, time 22316.39ms, mfu -100.00%
iter 1: loss nan, time 3249.91ms, mfu -100.00%
```
```bash
> python train.py config/nan.py --allow_fused=False
[...]
number of parameters: 11.39M
using fused AdamW: False
compiling the model... (takes a ~minute)
step 0: train loss 6.6010, val loss 6.6009
[2023-03-01 15:20:26,110] torch._inductor.utils: [WARNING] using triton random, expect difference from eager
/projects/nanoGPT_nan/venv/lib/python3.10/site-packages/torch/_functorch/aot_autograd.py:1251: UserWarning: Your compiler for AOTAutograd is returning a a function that doesn't take boxed arguments. Please wrap it with functorch.compile.make_boxed_func or handle the boxed arguments yourself. See https://github.com/pytorch/pytorch/pull/83137#issuecomment-1211320670 for rationale.
  warnings.warn(
iter 0: loss 6.6793, time 22286.14ms, mfu -100.00%
iter 1: loss 6.6811, time 3534.44ms, mfu -100.00%
iter 2: loss 6.6770, time 3365.06ms, mfu -100.00%
iter 3: loss 4.7208, time 3366.52ms, mfu -100.00%
iter 4: loss 3.0927, time 3365.22ms, mfu -100.00%
iter 5: loss 2.0694, time 3362.80ms, mfu 6.51%
iter 6: loss 1.4752, time 3360.81ms, mfu 6.51%
# continues to work fine
```

Hope this helps.

P.S.:

Can anyone explain what the warning means? Doesn't seem to be related to the issue, but is it a user error that should be fixed?
```
[2023-03-01 15:19:43,485] torch._inductor.utils: [WARNING] using triton random, expect difference from eager
/projects/venv/lib/python3.10/site-packages/torch/_functorch/aot_autograd.py:1251: UserWarning: Your compiler for AOTAutograd is returning a a function that doesn't take boxed arguments. Please wrap it with functorch.compile.make_boxed_func or handle the boxed arguments yourself. See https://github.com/pytorch/pytorch/pull/83137#issuecomment-1211320670 for rationale.
  warnings.warn(
```

### Versions

```
Collecting environment information...
PyTorch version: 2.0.0.dev20230220+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Arch Linux (x86_64)
GCC version: (GCC) 12.2.1 20230201
Clang version: 15.0.7
CMake version: version 3.25.0
Libc version: glibc-2.37

Python version: 3.10.9 (main, Dec 19 2022, 17:35:49) [GCC 12.2.0] (64-bit runtime)
Python platform: Linux-5.15.79.1-microsoft-standard-WSL2-x86_64-with-glibc2.37
Is CUDA available: True
CUDA runtime version: 11.8.89
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 2080 Ti
Nvidia driver version: 516.94
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   48 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          24
On-line CPU(s) list:             0-23
Vendor ID:                       AuthenticAMD
Model name:                      AMD Ryzen 9 5900X 12-Core Processor
CPU family:                      25
Model:                           33
Thread(s) per core:              2
Core(s) per socket:              12
Socket(s):                       1
Stepping:                        0
BogoMIPS:                        7400.01
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl tsc_reliable nonstop_tsc cpuid extd_apicid pni pclmulqdq ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm cmp_legacy cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw topoext ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 erms rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves clzero xsaveerptr arat umip vaes vpclmulqdq rdpid fsrm
Hypervisor vendor:               Microsoft
Virtualization type:             full
L1d cache:                       384 KiB (12 instances)
L1i cache:                       384 KiB (12 instances)
L2 cache:                        6 MiB (12 instances)
L3 cache:                        32 MiB (1 instance)
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Vulnerable
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Retpolines, IBPB conditional, IBRS_FW, STIBP conditional, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.24.1
[pip3] pytorch-triton==2.0.0+d54c04abe2
[pip3] torch==2.0.0.dev20230220+cu118
[pip3] torchaudio==2.0.0.dev20230223+cu118
[pip3] torchinfo==1.7.2
[pip3] torchsummary==1.5.1
[pip3] torchvision==0.15.0.dev20230223+cu118
[conda] Could not collect
```

cc @ezyang @gchanan @zou3519 @vincentqb @jbschlosser @albanD @janeyx99

### PR DESCRIPTION
**Pull request by crcrpar**
Fixes #95781.
The cause seems to be that the current implementation doesn't correctly pass `found_inf` when `grad_scale` is `None`. Therefore parameters can get mistakenly updated by gradients whose some elements are invalid, i.e. nan or inf.

Related #94060

I forgot about this wrong handling after #94344

cc @janeyx99

### CODE REVIEW

File: test/test_optim.py [hunk 1]
```diff
@@ -1647,6 +1649,7 @@ def test_functional_fused_optimizer_with_foundinf(self):
                     for _ in range(num_tensors)
                 ],
             )
+            self.assertEqual(params, prev_params)
```
Comment (janeyx99) [inline test/test_optim.py:1652]: If I understand properly, this test case tests that the the params are NOT mutated for functional optimizers, yes? If so, should we add another test that verifies the bug would not regress?

**Review by janeyx99 — COMMENTED** [review]
Thanks for looking into this! I have a general question about the test plan of this proposal--how can we assure the bug is fixed + prevent future regressions?

Could this also have to do with why the performance of fused adam has been reportedly less good? https://github.com/pytorch/pytorch/issues/90752#issuecomment-1427793057

File: test/test_optim.py (hunk 1 above)
Comment (crcrpar) [reply test/test_optim.py:1652]: The cause is that `found_inf` is not passed to the kernels when `grad_scale` is not set to the fused optimizers thus the parameters got updated using invalid gradients, leading to nan loss.
So one appropriate test is to make sure that the parameters are not updated after `optimizer.step` call if `found_inf` is `1` and there are some invalid gradients.

Comment (janeyx99) [reply test/test_optim.py:1652]: Ah lol my bad, I misinterpreted "functional" there. This test + fix make sense together.

Comment (janeyx99) [reply test/test_optim.py:1652]: Are there any docs on when grad_scale is set vs not set? In the issue, it looks like the bug repros had turned GradScaler to on since they're using float16, and this seems to catch a case when grad_scale is off. I'm going to approve since this fix seems like something we definitely want! I'd also like to understand AMP more and verify the bug would be completely fixed.

**Review by janeyx99 — APPROVED** [review]

File: test/test_optim.py (hunk 1 above)
Comment (crcrpar) [reply test/test_optim.py:1652]: when GradScaler is employed, 
- found_inf: needs to be set always
- grad_scale: needs to be set unless we manually call `GradScaler.unscale_`. A manual call of `GradScaler.unscale_` is sort of necessary to modify gradients, e.g. https://pytorch.org/docs/stable/generated/torch.nn.utils.clip_grad_norm_.html

**Comment by janeyx99** [comment]
@crcrpar Could you also cherry-pick this PR following the instructions in https://github.com/pytorch/pytorch/issues/97272 for the potential 2.0.1 release?
